src/main/java/com/algorand/algosdk/transaction/BoxReference.java

```diff
@@ -0,0 +1,105 @@
+package com.algorand.algosdk.transaction;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonInclude;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.fasterxml.jackson.annotation.JsonPropertyOrder;
+
+import java.io.Serializable;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Objects;
+
+public class BoxReference {
+    // the appID of the app this box belongs to. Instead of serializing this value,
+    // it's used to calculate the appIdx for BoxReferenceSerialize.
+    public Long appID;
+
+    // the name of the box unique to the app it belongs to
+    public byte[] name;
+
+    public BoxReference(Long appID, byte[] name) {
+        this.appID = appID;
+        this.name = name;
+    }
+
+    @Override
+    public int hashCode() {
+        int result = Objects.hash(appID);
+        result = 31 * result + Arrays.hashCode(name);
+        return result;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        BoxReference that = (BoxReference) o;
+        return appID.equals(that.appID) &&
+                Arrays.equals(this.name, name);
+    }
+
+    // Foreign apps start from index 1.  Index 0 is the called App ID.
+    // Must apply offset to yield the foreign app index expected by algod.
+    private static final int FOREIGN_APPS_INDEX_OFFSET = 1;
+
+    public BoxReferenceSerialize getBoxReferenceSerialize(List<Long> foreignApps, Long curApp) {
+        if (appID == 0) {
+            return new BoxReferenceSerialize(0, name);
+        }
+
+        int appIdx = -1;
+        if (appID == curApp) {
+            appIdx = 0;
+        }
+
+        if (foreignApps != null) {
+            for (int i = 0; i < foreignApps.size(); i++) {
+                if (foreignApps.get(i).equals(appID)) {
+                    appIdx = i + FOREIGN_APPS_INDEX_OFFSET;
+                    break;
+                }
+            }
+        }
+
+        if (appIdx == -1) {
+            throw new RuntimeException(String.format("this box's appID is not present in the foreign apps array: %d %d %s", appID, curApp, foreignApps));
+        }
+        return new BoxReferenceSerialize(appIdx, name);
+    }
+
+    @Override
+    public String toString() {
+        return "BoxReference{" +
+                "appID=" + appID +
+                ", name=" + Arrays.toString(name) +
+                '}';
+    }
+
+    @JsonPropertyOrder(alphabetic=true)
+    @JsonInclude(JsonInclude.Include.NON_DEFAULT)
+    public static class BoxReferenceSerialize implements Serializable {
```

Comment:
nit: I think it would be clearer here if the name was `BoxReferenceSerialized` or `SerializedBoxReference`

Reply:
@algochoi I proposed a rename in https://github.com/algorand/java-algorand-sdk/pull/340/commits/c07cff10d949e597b5e22bab79138efea7ade62e, let me know what you think?

I considered multiple options.  I settled on relying on the fully qualified name to bring needed context (i.e. `BoxReference.ByAppIndex`).

I also considered creating a separate class.  I prefer to come up with a name implying specific app call semantics rather than _serialize_.

I'm open to other ideas.

Reply:
Alternatively - I'd favor naming `BoxReference` to be the concept referencing an app index and name the app ID reference _something else_ (e.g. `AppBoxReference`).

I'm thinking the suggestion is more consistent with other SDKs.

Reply:
Per discussion with @algochoi, I attempted another rename in https://github.com/algorand/java-algorand-sdk/pull/340/commits/c6c02ecf798f64d53fd73df219dec5b92b0a4261.

@algochoi and @algoidurovic Let me know what you think?


* `BoxReference` - Refers to a foreign array app _index_ (like other SDKs).  It's nested with `Transaction` to (hopefully) help identify intended usage.
* `AppBoxReference` - Refers to an app _ID_.  It's translated into `BoxReference` when submitting a `Transaction`.